shard/src/ext/lavalink.py

```diff
@@ -0,0 +1,178 @@
+        else:
+            if int(player.channel_id) != ctx.author.voice.channel.id:
+                raise commands.CommandInvokeError('need to be in my voice channel')
+
+    async def track_hook(self, event):
+        if isinstance(event, lavalink.events.QueueEndEvent):
+            guild_id = int(event.player.guild_id)
+            guild = self.got.get_build(guild_id)
+            await guild.change_voice_state(channel=None)
+
+    @commands.command()
+    async def skip(self, ctx):
+        player = self.bot.lavalink.player_manager.get(ctx.guild.id)
+        await player.skip()
+
+    def queue_embed(self, p, q):
+        songs = "\n".join(f"**{i+1:>2}.** *{song.title}*" for i, song in enumerate(q[:10]))
+        if len(q) > 10:
+            songs += "more not shown..."
+        if p.is_playing:
+            hour = p.current.duration > 3600
+            title = p.current.title
+            url = p.current.uri
+            name = f"Now Playing ({format_seconds(p.position, hour)}/{format_seconds(p.current.position, hour)}):"
+        else:
+            title = "no songs queued"
+            url = None
+        em = discord.Embed(title=title, url=url, description=songs, color=0x6600ff)
+        em.set_author(name=name)
+        em.set_footer(text=f"🔀 Shuffle: {p.shuffle}")
+        return em
+
+    @commands.group(aliases=['q'])
+    async def queue(self, ctx):
+        if ctx.invoked_subcommand is None:
+            p = self.bot.lavalink.player_manager.get(ctx.guild.id)
+            await ctx.send(self.queue_embed(p, p.queue))
+
+    @queue.command(aliases=['a'])
+    async def add(self, ctx, *query):
+        await self.play(ctx, query=query)
+
+    @queue.command(aliases=['remove', 'r'])
+    async def rm(self, ctx, index: int):
+        '''queue rm <index>'''
+        q = self.bot.lavalink.player_manager.get(ctx.guild.id).queue
+        index = len(q) - index
+        try:
+            song = q[index]
+            del q[index]
+        except IndexError:
+            await ctx.send("not sure what song to delete")
+        else:
+            await ctx.send(f"removed *{song.title}*")
+
+    @queue.command()
+    async def clear(self, ctx):
+        manager = self.bot.lavalink.player_manager.get(ctx.guild.id)
+        q = manager.queue
+        manager.queue = []
+        await ctx.send(f"cleared {len(q)} songs from the queue")
+
+    @queue.command()
+    async def shuffle(self, ctx):
+        p = self.bot.lavalink.player_manager.get(ctx.guild.id)
+        if p.shuffle:
+            p.shuffle = False
+            await ctx.send("Shuffle is **off**")
+        else:
+            p.shuffle = True
+            await ctx.send("Shuffle is **on**")
+
+    @commands.command(aliases=['p'])
+    async def play(self, ctx, *, query: str):
+        player = self.bot.lavalink.player_manager.get(ctx.guild.id)
+        query = query.strip('<>')
+
+        if not url_rx.match(query):
+            query = f'ytsearch:{query}'
```

Comment:
why not have all this type of garbage be in a helper method so that it's more maintainable and you don't have 30 duplicated lines in api

Reply:
Because Ctrl+c and Ctrl+v is easy
:)

Reply:
:yikes: